### DISCUSSION TITLE
Cypress component testing Vue 3

### DISCUSSION THREAD

**Discussion by 1beb** [cypress-io/cypress#16714] [Component Testing]
When attempting to install the dependencies as described in the related [blogpost](https://www.cypress.io/blog/2021/04/06/getting-start-with-cypress-component-testing-vue-2-3/) I was unable to install the dependencies: 

I tried this with an existing project and also a brand new Vue 3 project with nothing else with a factory fresh install of `vue cli`: 

```
$ sudo npm install -g @vue/cli
...
$ vue create test-project
$ cd test-project
$ npm install @cypress/vue@next --dev
npm WARN install Usage of the `--dev` option is deprecated. Use `--include=dev` instead.
npm ERR! code ERESOLVE
npm ERR! Cannot destructure property 'name' of 'node' as it is null.
npm ERR! A complete log of this run can be found in:
npm ERR! /home/user/.npm/_logs/2021-05-27T21_29_21_247Z-debug.log
```
Note: I have 3 colleagues, 2 of which are on Ubuntu and expeienced the same error. One of which is Windows and did not experience the error. 

Windows user with success npm version 6.14.11
Ubuntu users with failure npm version 7.7.6

**Comment by lmiller1990** [comment]
I tried this on Mac OS and it's working as expected for me. 🤔

**Reply by 1beb** [reply]
Yeah, it's definitely a platform issue with dependency resolution: 

```
npm install @cypress/vue@next --dev --force
npm WARN using --force Recommended protections disabled.
npm WARN install Usage of the `--dev` option is deprecated. Use `--include=dev` instead.
npm WARN ERESOLVE overriding peer dependency
npm WARN While resolving: test-project@0.1.0
npm WARN Found: html-webpack-plugin@5.3.1
npm WARN node_modules/html-webpack-plugin
npm WARN   peer html-webpack-plugin@">=4" from @cypress/webpack-dev-server@1.3.1
npm WARN   node_modules/@cypress/webpack-dev-server
npm WARN     peerOptional @cypress/webpack-dev-server@"*" from @cypress/vue@3.0.1
npm WARN     node_modules/@cypress/vue
npm WARN 
npm WARN Could not resolve dependency:
npm WARN peer html-webpack-plugin@">=4" from @cypress/webpack-dev-server@1.3.1
npm WARN node_modules/@cypress/webpack-dev-server
npm WARN   peerOptional @cypress/webpack-dev-server@"*" from @cypress/vue@3.0.1
npm WARN   node_modules/@cypress/vue
```

**Reply by flotwig** [reply]
Fails for me on Ubuntu 20.10 with npm 7.15.0 with a different error:

```
npm WARN install Usage of the `--dev` option is deprecated. Use `--include=dev` instead.
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR! 
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! @cypress/vue@"3.0.1" from the root project
npm ERR! 
npm ERR! Conflicting peer dependency: @cypress/webpack-dev-server@1.3.1
npm ERR! node_modules/@cypress/webpack-dev-server
npm ERR!   peerOptional @cypress/webpack-dev-server@"*" from @cypress/vue@3.0.1
npm ERR!   node_modules/@cypress/vue
npm ERR!     @cypress/vue@"3.0.1" from the root project
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force, or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
npm ERR! 
npm ERR! See /home/flotwig/.npm/eresolve-report.txt for a full report.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/flotwig/.npm/_logs/2021-05-28T15_38_04_025Z-debug.log
```

[eresolve-report.txt](https://github.com/cypress-io/cypress/files/6561857/eresolve-report.txt)
[2021-05-28T15_38_04_025Z-debug.log](https://github.com/cypress-io/cypress/files/6561861/2021-05-28T15_38_04_025Z-debug.log)

**Reply by lmiller1990** [reply]
I tried using docker to replicate, it looks like you must be on an real linux machine to encounter this. I was using 

```
root@fdec2ff3b9a0:/home/person/test-project# npm -v
7.15.1

root@fdec2ff3b9a0:/home/person/test-project# node -v
v14.16.0

root@fdec2ff3b9a0:/home/person/test-project# npm install @cypress/vue@next --dev

npm WARN install Usage of the `--dev` option is deprecated. Use `--include=dev` instead.
changed 308 packages, and audited 1274 packages in 8s
1 package is looking for funding
  run `npm fund` for details
48 moderate severity vulnerabilities
To address issues that do not require attention, run:
  npm audit fix
To address all issues (including breaking changes), run:
  npm audit fix --force
Run `npm audit` for details.
root@fdec2ff3b9a0:/home/person/test-project#
```

I will fire up my linux desktop and try to debug this.

**Reply by GustafHultgren** [reply]
Is there a fix for this yet? I'm on windows 10, npm 7.10.0 and cypress 8.3.0 trying to install @cypress/vue@next.
I get a dependency resolution error from npm. I guess @cypress/vue@next has an optional dependency on @cypress/webpack-dev-server@"*" whereas @cypress/ has a dependency on @cypress/webpack-dev-server@"1.4.0". 

In npm 7 this produces a conflict.

```sh
npm ERR! Conflicting peer dependency: @cypress/webpack-dev-server@1.4.0
npm ERR! node_modules/@cypress/webpack-dev-server
npm ERR!   peerOptional @cypress/webpack-dev-server@"*" from @cypress/vue@3.0.3
npm ERR!   node_modules/@cypress/vue
npm ERR!     dev @cypress/vue@"3.0.3" from the root project
```

There is an issue with npm 7+ as well, which might be the root of the problem I guess.
https://stackoverflow.com/questions/67185714/how-do-i-read-npm-conflicting-peer-dependency-error-messages

**Reply by lmiller1990** [reply]
Thanks for the info - this helps a lot, I guess this is just npm getting stricter. I think we can probably fix this just by tweaking the dependencies.

I cannot look at this right now, maybe early next week, but happy to accept a PR if anyone is has a moment. If you make a PR, please fork from master - it'll be released immediately via semantic release once we merge it.

**Reply by ZachJW34** [reply]
I was able to reproduce this issue across most of the npm versions listed above. The error I received was:
```
zachjw@Zacharys-MacBook-Pro test-project % npm install cypress @cypress/vue@next                                  
npm ERR! code ERESOLVE
npm ERR! ERESOLVE could not resolve
npm ERR! 
npm ERR! While resolving: test-project@0.1.0
npm ERR! Found: html-webpack-plugin@3.2.0
npm ERR! node_modules/html-webpack-plugin
npm ERR!   html-webpack-plugin@"^3.2.0" from @vue/cli-service@4.5.13
npm ERR!   node_modules/@vue/cli-service
npm ERR!     peer @vue/cli-service@"^3.0.0 || ^4.0.0-0" from @vue/cli-plugin-babel@4.5.13
npm ERR!     node_modules/@vue/cli-plugin-babel
npm ERR!       dev @vue/cli-plugin-babel@"~4.5.0" from the root project
npm ERR!     peer @vue/cli-service@"^3.0.0 || ^4.0.0-0" from @vue/cli-plugin-eslint@4.5.13
npm ERR!     node_modules/@vue/cli-plugin-eslint
npm ERR!       dev @vue/cli-plugin-eslint@"~4.5.0" from the root project
npm ERR!     3 more (@vue/cli-plugin-router, @vue/cli-plugin-vuex, the root project)
npm ERR!   peer html-webpack-plugin@">=2.26.0" from @vue/preload-webpack-plugin@1.1.2
npm ERR!   node_modules/@vue/preload-webpack-plugin
npm ERR!     @vue/preload-webpack-plugin@"^1.1.0" from @vue/cli-service@4.5.13
npm ERR!     node_modules/@vue/cli-service
npm ERR!       peer @vue/cli-service@"^3.0.0 || ^4.0.0-0" from @vue/cli-plugin-babel@4.5.13
npm ERR!       node_modules/@vue/cli-plugin-babel
npm ERR!         dev @vue/cli-plugin-babel@"~4.5.0" from the root project
npm ERR!       4 more (@vue/cli-plugin-eslint, @vue/cli-plugin-router, ...)
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! @cypress/vue@"3.0.3" from the root project
npm ERR! 
npm ERR! Conflicting peer dependency: html-webpack-plugin@5.3.2
npm ERR! node_modules/html-webpack-plugin
npm ERR!   peer html-webpack-plugin@">=4" from @cypress/webpack-dev-server@1.4.0
npm ERR!   node_modules/@cypress/webpack-dev-server
npm ERR!     peerOptional @cypress/webpack-dev-server@"*" from @cypress/vue@3.0.3
npm ERR!     node_modules/@cypress/vue
npm ERR!       @cypress/vue@"3.0.3" from the root project
```
This error makes sense, as `@cypress/webpack-dev-server` requires `html-webpack-plugin >= 4` but `@vue/cli-service@4.5.13` uses `html-webpack-plugin = ^3.2.0`. Not quite sure of the resolution for this yet, we can make `html-webpack-plugin` a direct dependency rather than peer, but I'll have to return to this issue later.

**Reply by lmiller1990** [reply]
We have had many issues around `html-webpack-plugin`. 

Ref: https://github.com/cypress-io/cypress/pull/16108
Ref: https://github.com/cypress-io/cypress/pull/16388

Basically `html-webpack-plugin@3.2` is two major versions behind. Users need to be on 4.x (for webpack 4) or 5.x (for webpack 5). Having `html-webpack-plugin@3` installed actually breaks `@cypress/webpack-dev-server`. We actually remove several webpack plugins that cause problems, [see here](https://github.com/cypress-io/cypress/blob/develop/npm/webpack-dev-server/src/makeWebpackConfig.ts#L70). We then require the user either bring html-webpack-plugin 4.x or 5.x - basically, we do not support 3.x, since it's so out of date. Many projects are still using 3.x, but likely can be updated - both CRA and Vue CLI are on 4.x or 5.x now, for Webpack 4.x and 5.x respectively.

It looks like `@vue/cli-service` is now using `html-webpack-plugin@5.x`: https://github.com/vuejs/vue-cli/blob/c1bf3d874692637648a5cee39c8ea27126272d35/packages/%40vue/cli-service/package.json#L60. It also moved to 4.x around 9 months ago: https://github.com/vuejs/vue-cli/blame/f9d53c75d0e6c8a0cc1ff4ab7a0634e9c6fdce06/packages/@vue/cli-service/package.json#L59

@1beb are you able to try updating your Vue CLI dependencies? That might solve your problem.

**Reply by 1beb** [reply]
HI @lmiller1990, I'm going to be working on this again this week. Apologies for the hiatus. I will update to let you know if this resolved the issue.

**Reply by richardshergold** [reply]
I'm getting this error now trying to install in my Vue 3 project - is there a solution?  I'm getting:

```
~/projects/my-project [cypress] > npm install @cypress/vue@next --include=dev
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR!
npm ERR!
npm ERR! Could not resolve dependency:
npm ERR! @cypress/vue@"3.0.5" from the root project
npm ERR!
npm ERR! Conflicting peer dependency: @cypress/webpack-dev-server@1.7.0
npm ERR! node_modules/@cypress/webpack-dev-server
npm ERR!   peerOptional @cypress/webpack-dev-server@"*" from @cypress/vue@3.0.5
npm ERR!   node_modules/@cypress/vue
npm ERR!     @cypress/vue@"3.0.5" from the root project
npm ERR!
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force, or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
```

**Reply by lmiller1990** [reply]
Stupid question: what if you do `--force` like it says? Not ideal, but good to know if that will let you proceed and use Cypress with Vue 3.

What npm version are you on? Looks like we need to look into our dependencies again and make sure everything is correct.

**Reply by richardshergold** [reply]
Not a stupid question!  I've not tried --force yet.  I will try it later and report back (diverted onto something else right now)  

npm is 7.17.0    

(I did create a separate issue https://github.com/cypress-io/cypress/issues/19106)
